indi-orion-ssg3/orion_ssg3_ccd.cpp

```diff
@@ -0,0 +1,580 @@
+
+    timesince = (double)(now.tv_sec * 1000.0 + now.tv_usec / 1000) -
+                (double)(ExpStart.tv_sec * 1000.0 + ExpStart.tv_usec / 1000);
+    timesince = timesince / 1000;
+
+    timeleft = ExposureRequest - timesince;
+    return timeleft;
+}
+
+/**
+ * Client is asking us to set a new number
+ */
+bool SSG3CCD::ISNewNumber(const char *dev, const char *name, double values[], char *names[], int n)
+{
+    int rc;
+
+    if (!strcmp(dev, getDeviceName()))
+    {
+        if (GainNP.isNameMatch(name))
+        {
+            if (GainNP.update(values, names, n) == false)
+            {
+                GainNP.setState(IPS_ALERT);
+                GainNP.apply();
+                return true;
+            }
+            LOGF_INFO("Setting gain to %.f", GainNP[0].getValue());
+            rc = orion_ssg3_set_gain(&ssg3, GainNP[0].getValue());
+            if (!rc) {
+                GainNP.setState(IPS_OK);
+            } else {
+                GainNP.setState(IPS_ALERT);
+            }
+            GainNP.apply();
+            return true;
+        }
+
+        if (OffsetNP.isNameMatch(name))
+        {
+            if (OffsetNP.update(values, names, n) == false)
+            {
+                OffsetNP.setState(IPS_ALERT);
+                OffsetNP.apply();
+                return true;
+            }
+            LOGF_INFO("Setting offset to %.f", OffsetNP[0].getValue());
+            rc = orion_ssg3_set_offset(&ssg3, OffsetNP[0].getValue());
+            if (!rc) {
+                OffsetNP.setState(IPS_OK);
+            } else {
+                OffsetNP.setState(IPS_ALERT);
+            }
+            OffsetNP.apply();
+            return true;
+        }
+    }
+
+    // If we didn't process anything above, let the parent handle it.
+    return INDI::CCD::ISNewNumber(dev, name, values, names, n);
+}
+
+/**
+ * Client is asking us to set a new switch
+ */
+bool SSG3CCD::ISNewSwitch(const char *dev, const char *name, ISState *states, char *names[], int n)
+{
+    if (!strcmp(dev, getDeviceName()))
+    {
+        if (CoolerSP.isNameMatch(name))
+        {
+            if (CoolerSP.update(states, names, n) == false)
+            {
+                CoolerSP.setState(IPS_ALERT);
+                CoolerSP.apply();
+                return true;
+            }
+            IUUpdateSwitch(&CoolerSP, states, names, n);
+            CoolerSP.setState(IPS_OK);
+            IDSetSwitch(&CoolerSP, nullptr);
+
+            if (CoolerSP[0].getState() == ISS_OFF) {
+                activateCooler(false);
+            } else {
+                activateCooler(true);
+            }
+
+            return true;
+        }
+    }
+    return INDI::CCD::ISNewSwitch(dev, name, states, names, n);
+}
+
+/**
+ * Main device loop. We check for exposure progress
+ */
+void SSG3CCD::TimerHit()
+{
+    long timeleft;
+
+    if (isConnected() == false) {
+        return;
+    }
+
+    if (InExposure) {
+        timeleft = CalcTimeLeft();
```

Comment:
No need to use this function, use INDI::ElapsedTimer which is better than this method.